packets.go

```diff
@@ -571,9 +571,19 @@ func (mc *mysqlConn) handleErrorPacket(data []byte) error {
 	// Error Number [16 bit uint]
 	errno := binary.LittleEndian.Uint16(data[1:3])
 
+	pos := 3
+
+	// SQL State [optional: # + 5bytes string]
+	var sqlstate string
+	if data[3] == 0x23 {
+		sqlstate = string(data[4 : 4+5])
+		pos = 9
+	}
+
 	// 1792: ER_CANT_EXECUTE_IN_READ_ONLY_TRANSACTION
 	// 1290: ER_OPTION_PREVENTS_STATEMENT (returned by Aurora during failover)
-	if (errno == 1792 || errno == 1290) && mc.cfg.RejectReadOnly {
+	// SQLSTATE 45000: user-defined exception
+	if (errno == 1792 || errno == 1290) && sqlstate != "45000" && mc.cfg.RejectReadOnly {
```

Comment:
I am not sure about this is good idea.
User explicitly chose errno 1792 or 1290, and set `RejectReadOnly`. Why we need to check this?
If user want to cause reconnection by custom error, they need to chose sqlstate other than '45000'?

How about other places? Should all places checking errno check `sqlstate != "45000"` too? Or is there any reason this place is the only special case?

Reply:
It is a bug not to check the SQLState. Using stored routines and the `SIGNAL` statement should not lead to connection aborts because the library ignored the SQLState. I don't think it is unreasonable to use stored routines and wanting to
RejectReadOnly connections. We intent to use the library in that way.

---

As far as I can tell this is the only place where `errno` is used. I am a new user of this library though.

---

This could as well be written to check for the SQLStates documented in the MySQL manual. For 5.7 and 8.0. These are:

* 1290: HY000
* 1792: 25006

I chose to exclude 45000 since it has less chance to cause trouble when an implementation does not set SQLState or sends a SQLState not mentioned above.

45000 is the only SQLState [documented](https://dev.mysql.com/doc/refman/8.0/en/signal.html) to be reserved for the user i know of:

> To signal a generic SQLSTATE value, use '45000', which means “unhandled user-defined exception.”

Reply:
> Using stored routines and the `SIGNAL` statement should not lead to connection aborts because the library ignored the SQLState.

Even the SIGNAL abused existing MySQL errno?
Would you give us some official guideline about using system defined errno in SIGNAL?

Additionally, do you really want to use the `RejectReadOnly` option?
If so, I think you should avoid using system errno in user defined error, at least 1290.

As far as I know, ProxySQL and MariaDB Connector/J has a feature similar to RejectReadOnly.
They check only errno==1290, but don't check SQLSTATE.

ProxySQL:

https://github.com/sysown/proxysql/blob/91d99aacde5133c0882959091830e7f64737bd74/lib/MySQL_Session.cpp#L795-L798

MariaDB Connector/J:

https://github.com/mariadb-corporation/mariadb-connector-j/blob/7e4650ddb17d17d46c7078f90f8c7510ce1f9d06/src/main/java/org/mariadb/jdbc/internal/failover/FailoverProxy.java#L202-L209

AWS RDS Proxy has similar feature (Aurora fast failover), but it is not open sourced. So I am not sure they check SQLState.

Reply:
I could not find any official documentation proofing me right. Reading the documentation of [SIGNAL](https://dev.mysql.com/doc/refman/8.0/en/signal.html) and [RESIGNAL](https://dev.mysql.com/doc/refman/8.0/en/resignal.html) in my opinion suggest that i am right. In no word is suggested that attention has to be paid to not collide errnos. Further, SQLState 45000 is explicitly reserved for the user.

[RESIGNAL](https://dev.mysql.com/doc/refman/8.0/en/resignal.html#resignal-with-condition) uses SQLSTATE 45000 and MYSQL_ERRNO = 5 in an example. Interpreted without the SQLState this would mean
> Error number: 5; Symbol: [EE_OUTOFMEMORY](https://dev.mysql.com/doc/mysql-errors/8.0/en/global-error-reference.html#error_ee_outofmemory);

Is the user now expected to free memory?

If you want to `SIGNAL SQLSTATE 'HY000' SET MYSQL_ERRNO = 1290;`, you can just do so. All existing MySQL errnos have a documented SQLState. 
I don't think you should ever do this. It is not up to the user to raise implementation errors. The documentation only uses '45000' to raise user exceptions and '01000' to record warnings.

`SIGNAL` is not abusing existing MySQL errnos since it uses a different SQLState explicitly reserved for the user.

---

We would have to check 1500+ procedures for every new MySQL server version to make sure there are no collisions. Then update the procedures and all applications using them and redeploy the applications. This would render the `SIGNAL` statement useless. I believe this is the reason for reserving a SQLState for the user.

I sure can live without the RejectReadOnly feature. If this is keeping `MySQLError` from being extended with the SQLState I will gladly drop it.

Reply:
> I sure can live without the RejectReadOnly feature. If this is keeping MySQLError from being extended with the SQLState I will gladly drop it.

Please do it. ProxySQL is the most major OSS product which provides fast failover using read-only.
I don't want to add more random conditions that is not implemented in ProxySQL.

In other words, if ProxySQL admits checking only errno==1290 was bug and change its behavior, I am happy to follow it.

Reply:
Removed SQLState check.